File: sdk/python/src/marmot/auth/exchange.py
"""RFC 8693 token exchange against Marmot's /oauth/token endpoint."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
from typing import TYPE_CHECKING

import httpx

from marmot.errors import AuthError, ServerError

if TYPE_CHECKING:
    from marmot.auth import Credential

GRANT_TYPE = "urn:ietf:params:oauth:grant-type:token-exchange"
TOKEN_TYPE_ID_TOKEN = "urn:ietf:params:oauth:token-type:id_token"
TOKEN_TYPE_ACCESS_TOKEN = "urn:ietf:params:oauth:token-type:access_token"

_DEFAULT_TIMEOUT = 10.0


def exchange(
    *,
    base_url: str,
    subject_token: str,
    subject_token_type: str,
    source_name: str,
    timeout: float = _DEFAULT_TIMEOUT,
    client: httpx.Client | None = None,
) -> Credential:
    """Exchange a workload-identity token for a Marmot session token.

    Returns a :class:`marmot.auth.Credential` whose ``refresh`` callback
    re-runs the exchange — useful for long-lived agents.
    """
    from marmot.auth import Credential  # avoid circular import

    def _do_exchange() -> Credential:
        url = base_url.rstrip("/") + "/oauth/token"
        data = {
            "grant_type": GRANT_TYPE,
            "subject_token": subject_token,
            "subject_token_type": subject_token_type,
        }

        owns_client = client is None
        c = client or httpx.Client(timeout=timeout)
        try:
            resp = c.post(url, data=data)
        finally:
            if owns_client:
                c.close()

        if resp.status_code == 400:
            raise AuthError(_oauth_error(resp))
        if resp.status_code == 401:
            raise AuthError(
                f"server rejected workload-identity token from {source_name}: {_oauth_error(resp)}"
            )
        if resp.status_code >= 500:
            raise ServerError(
                f"token exchange failed: HTTP {resp.status_code}", status_code=resp.status_code
            )
        if resp.status_code != 200:
            raise ServerError(
                f"unexpected response from /oauth/token: HTTP {resp.status_code}",
                status_code=resp.status_code,
            )

        body = resp.json()
        access = body.get("access_token")
        if not isinstance(access, str) or not access:
            raise ServerError("token exchange response missing access_token")

        expires_in = body.get("expires_in")
        expires_at: datetime | None = None
        if isinstance(expires_in, int) and expires_in > 0:
            expires_at = datetime.now(timezone.utc) + timedelta(seconds=expires_in)

        return Credential(
            token=access,
            scheme="Bearer",
            expires_at=expires_at,
            refresh=_do_exchange,
            source=f"token exchange via {source_name}",
        )

    return _do_exchange()


def _oauth_error(resp: httpx.Response) -> str:
    """Format an RFC 6749 error response."""
    try:
        body = resp.json()
    except ValueError:
        return resp.text or f"HTTP {resp.status_code}"

    err = body.get("error", "")
    desc = body.get("error_description", "")
    if err and desc:
        return f"{err}: {desc}"
    return err or desc or resp.text or f"HTTP {resp.status_code}"

File: sdk/ts/src/auth/exchange.ts
/** RFC 8693 token exchange against Marmot's /oauth/token endpoint. */

import { AuthError, ServerError } from "../errors.js";
import type { Credential } from "./index.js";

export const GRANT_TYPE = "urn:ietf:params:oauth:grant-type:token-exchange";
export const TOKEN_TYPE_ID_TOKEN = "urn:ietf:params:oauth:token-type:id_token";
export const TOKEN_TYPE_ACCESS_TOKEN = "urn:ietf:params:oauth:token-type:access_token";

export interface ExchangeArgs {
  baseUrl: string;
  subjectToken: string;
  subjectTokenType: string;
  sourceName: string;
  fetchImpl?: typeof fetch;
  timeoutMs?: number;
}

export async function exchange(args: ExchangeArgs): Promise<Credential> {
  const doExchange = async (): Promise<Credential> => {
    const url = `${args.baseUrl.replace(/\/$/, "")}/oauth/token`;
    const body = new URLSearchParams({
      grant_type: GRANT_TYPE,
      subject_token: args.subjectToken,
      subject_token_type: args.subjectTokenType,
    });

    const ac = new AbortController();
    const timer = setTimeout(() => ac.abort(), args.timeoutMs ?? 10_000);

    let resp: Response;
    try {
      resp = await (args.fetchImpl ?? fetch)(url, {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body,
        signal: ac.signal,
      });
    } finally {
      clearTimeout(timer);
    }

    if (resp.status === 400) throw new AuthError(await oauthErrorMessage(resp));
    if (resp.status === 401) {
      throw new AuthError(
        `server rejected workload-identity token from ${args.sourceName}: ${await oauthErrorMessage(resp)}`,
      );
    }
    if (resp.status >= 500) {
      throw new ServerError(`token exchange failed: HTTP ${resp.status}`, resp.status);
    }
    if (resp.status !== 200) {
      throw new ServerError(
        `unexpected response from /oauth/token: HTTP ${resp.status}`,
        resp.status,
      );
    }

    const json = (await resp.json()) as Record<string, unknown>;
    const accessToken = json.access_token;
    if (typeof accessToken !== "string" || !accessToken) {
      throw new ServerError("token exchange response missing access_token");
    }

    const expiresIn = typeof json.expires_in === "number" ? json.expires_in : 0;
    const expiresAt = expiresIn > 0 ? new Date(Date.now() + expiresIn * 1000) : undefined;

    return {
      token: accessToken,
      scheme: "Bearer",
      expiresAt,
      refresh: doExchange,
      source: `token exchange via ${args.sourceName}`,
    };
  };

  return doExchange();
}

async function oauthErrorMessage(resp: Response): Promise<string> {
  try {
    const body = (await resp.json()) as Record<string, unknown>;
    const err = typeof body.error === "string" ? body.error : "";
    const desc = typeof body.error_description === "string" ? body.error_description : "";
    if (err && desc) return `${err}: ${desc}`;
    return err || desc || `HTTP ${resp.status}`;
  } catch {
    try {
      return (await resp.text()) || `HTTP ${resp.status}`;
    } catch {
      return `HTTP ${resp.status}`;
    }
  }
}
